### PR TITLE
build: adds integration tests tooling

### PR DESCRIPTION
**Pull request by tdstein**
## Overview

Adds integration test infrastructure. The architecture is designed to provide fast execution across the Connect version grid within a developer environment. On my machine, the full test grid is completed in 30 to 45 seconds.

## Quick start

Run `make help` from the `integration` directory.

Run `make latest` to run the test suite against 2024.05.0.

## Architecture summary

A new Makefile is defined, which provides rules for executing the integration test suite against Connect. The Connect version is parameterized across `Makefile` and `docker-compose.yaml` via the `CONNECT_VERSION` environment variable. The Connect environment is configured for "bootstrapping" via `CONNECT_BOOTSTRAP_ENABLED`. The existing `rsconnect` implementation obtains the `CONNECT_API_KEY ` in `make test`. Ideally, this will be replaced in the future with an SDK implementation.

- `Makefile`: defines commands for integration test invocation.
- `docker-compose.yaml`: defines environment integration testing and invokes test suite
- `.github/workflows/ci.yaml`: adds integration-test job that runs test suite in a matrix.

### CODE REVIEW

File: src/posit/connect/client.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 from __future__ import annotations
 
+import requests
```
Comment (nealrichardson) [inline src/posit/connect/client.py:5]: What is this for?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -39,6 +39,36 @@ jobs:
...
+          - 2023.06.0
+          - 2023.05.0
+          - 2023.01.1
+          - 2023.01.0
+          - 2022.12.0
+          - 2022.11.0
+    steps:
+    - uses: actions/checkout@v4
+    - uses: docker/setup-buildx-action@v3
+    - uses: docker/setup-qemu-action@v3
```
Comment (nealrichardson) [inline .github/workflows/ci.yaml:66]: What do we need qemu for? Everything in this stack is x86 right?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
+FROM python:3
```
Comment (nealrichardson) [inline Dockerfile:1]: What is this Dockerfile used for?

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -39,6 +39,36 @@ jobs:
...
+          - 2023.12.0
+          - 2023.10.0
+          - 2023.09.0
+          - 2023.07.0
+          - 2023.06.0
+          - 2023.05.0
+          - 2023.01.1
+          - 2023.01.0
+          - 2022.12.0
+          - 2022.11.0
```
[the remark this answers is not in the document]
Comment (tdstein) [reply .github/workflows/ci.yaml:62]: They all complete around the same time when run concurrently. I think it should be whatever our support window is.

File: integration/docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    # Run integration test suite.
+    #
+    # Target is relative to the ./integration directory, not the project root
+    # directory. The execution base directory is determined by the 'WORKDIR'
+    # in the Dockerfile.
+    command: make -C ./integration test
+    environment:
+        - CONNECT_BOOTSTRAP_SECRETKEY=${CONNECT_BOOTSTRAP_SECRETKEY}
+        # Port 3939 is the default port for Connect
+        - CONNECT_SERVER=http://connect:3939
```
[the remark this answers is not in the document]
Comment (tdstein) [reply integration/docker-compose.yaml:13]: The hostname is the relative service name. So https://github.com/posit-dev/posit-sdk-py/blob/f8e38a79a7a33cae99f56c3431fe90170a281c2b/integration/docker-compose.yaml#L22

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (tdstein) [reply .github/workflows/ci.yaml:66]: Oh, good catch. I'll try removing it.

File: Dockerfile (hunk 3 above)
Comment (tdstein) [reply Dockerfile:1]: This is the test runner for the "tests" service in the compose file. https://github.com/posit-dev/posit-sdk-py/blob/f8e38a79a7a33cae99f56c3431fe90170a281c2b/integration/docker-compose.yaml#L3

**Pull request merged**